Specifications/Language/5_Grammar/QSharpParser.g4

```diff
@@ -0,0 +1,257 @@
+
+qualifiedName : Identifier ('.' Identifier)*;
+
+namespaceElement
+    : openDirective
+    | typeDeclaration
+    | callableDeclaration
+    ;
+
+// Open Directive
+
+openDirective : 'open' qualifiedName ('as' qualifiedName)? ';';
+
+// Declaration
+
+attribute : '@' expression;
+
+access : 'internal';
+
+declarationPrefix : attribute* access?;
+
+// Type Declaration
+
+typeDeclaration : declarationPrefix 'newtype' Identifier '=' underlyingType ';';
+
+underlyingType
+    : typeDeclarationTuple
+    | type
+    ;
+
+typeDeclarationTuple : '(' (typeTupleItem (',' typeTupleItem)*)? ')';
+
+typeTupleItem
+    : namedItem
+    | underlyingType
+    ;
+
+namedItem : Identifier ':' type;
+
+// Callable Declaration
+
+callableDeclaration
+    : declarationPrefix ('function' | 'operation')
+      Identifier typeParameterBinding? parameterTuple
+      ':' type characteristics?
+      callableBody
+    ;
+
+typeParameterBinding : '<' (TypeParameter (',' TypeParameter)*)? '>';
+
+parameterTuple : '(' (parameter (',' parameter)*)? ')';
+
+parameter
+    : namedItem
+    | parameterTuple
+    ;
+
+characteristics : 'is' characteristicsExpression;
+
+characteristicsExpression
+    : 'Adj' # AdjointCharacteristic
+    | 'Ctl' # ControlledCharacteristic
+    | '(' characteristicsExpression ')' # CharacteristicsGroup
+    | characteristicsExpression '*' characteristicsExpression # IntersectCharacteristics
+    | characteristicsExpression '+' characteristicsExpression # UnionCharacteristics
+    ;
+
+callableBody
+    : BraceLeft specialization* BraceRight # CallableSpecialization
+    | scope # CallableScope
+    ;
+
+specialization : specializationName+ specializationGenerator;
+
+specializationName
+    : 'body'
+    | 'adjoint'
+    | 'controlled'
+    ;
+
+specializationGenerator
+    : 'auto' ';' # AutoGenerator
+    | 'self' ';' # SelfGenerator
+    | 'invert' ';' # InvertGenerator
+    | 'distribute' ';' # DistributeGenerator
+    | 'intrinsic' ';' # IntrinsicGenerator
+    | providedSpecialization # ProvidedGenerator
+    ;
+
+providedSpecialization : specializationParameterTuple? scope;
+
+specializationParameterTuple : '(' (specializationParameter (',' specializationParameter)*)? ')';
+
+specializationParameter
+    : Identifier # SpecializationNamedParameter
+    | '...' # SpecializationImplicitParameter
+    ;
+
+// Type
+
+type
+    : '_' # MissingType
+    | TypeParameter # TypeParameter
+    | 'BigInt' # BigIntType
+    | 'Bool' # BoolType
+    | 'Double' # DoubleType
+    | 'Int' # IntType
+    | 'Pauli' # PauliType
+    | 'Qubit' # QubitType
+    | 'Range' # RangeType
+    | 'Result' # ResultType
+    | 'String' # StringType
+    | 'Unit' # UnitType
+    | qualifiedName # UserDefinedType
+    | '(' (type (',' type)* ','?)? ')' # TupleType
+    | '(' arrowType characteristics? ')' # CallableType
+    | type '[' ']' # ArrayType
+    ;
+
+arrowType
+    : '(' type ('->' | '=>') type ')'
+    | type ('->' | '=>') type
```

Comment:
a bit forward looking? ;)

Reply:
The grammar still requires parentheses around callable types, actually. This particular rule is for the optional parentheses *inside* the callable type:

* `((Qubit => Unit) is Adj)`
* `(Qubit => Unit is Adj)`

Reply:
Ah, sorry, I didn't read carefully enough. We should consider dropping support for inner parenthesis along with the other parentheses we are intending to remove.